lib/postprocessing/image.js

```diff
@@ -192,7 +199,47 @@ async function read(infile) {
     return gm(`${infile}[0]`);
 }
 
-async function applyOperations(img, instructions) {
+function getImageSize(metadata) {
+    // The orientation corresponds to the TIFF orientation tag which ranges fromn 1 to 8
+    // Values 5 to 8 require a swap of width and height
+    // See https://www.awaresystems.be/imaging/tiff/tifftags/orientation.html or
+    // https://www.adobe.io/content/dam/udp/en/open/standards/tiff/TIFF6.pdf
+    if(CHANGE_ORIENTATION_LIST.includes(metadata.orientation)) {
+        return new Dimensions(metadata.pageGeometry.height, metadata.pageGeometry.width);
+    } else {
+        return new Dimensions(metadata.pageGeometry);
+    }
+}
+async function applyWatemark(img, rendition, directories) {
+    const instructions = rendition.instructions;
+    const targetBoundingBox = new Dimensions(instructions.width, instructions.height);
+
+    // TODO drop old watermark.widthPercent once no longer used by AEM
+    const watermarkScale = instructions.watermark.scale || (instructions.watermark.widthPercent / 100.0) || 1.0;
+
+    // determine image dimensions
+    const imageSize = getImageSize(rendition.intermediateMetadata);
+
+    // TODO drop old watermark.watermarkContent once no longer used by AEM
+    const watermarkSource = instructions.watermark.image || instructions.watermark.watermarkContent;
+    // download watermark
+    if(!watermarkSource) {
+        throw new GenericError(`post-processing: Missing watermark.image in instructions : ${JSON.stringify(rendition.instructions)}`);
+    }
+    const asset = await getAsset(watermarkSource, directories.in, WATERMARK_NAME);
+    // TODO: infuture should we enforce that watermark is a png ?
+    // TODO: infuture handle same watermark png across multiple renditions
```

Comment:
(Local) caching for watermarks?

Reply:
yes. `getAsset()` should handle that. but it is a bit more involved so we skipped it for now. we'll see how often customers will ask for multiple renditions with watermark in one request, and whether AEM will send the same or different presigned urls in this case.